README.md

```diff
@@ -1,31 +1,784 @@
+}
+```
+
+`RequestPerformer` duty is to communicate with the server (perform requests). This is done in `perform` method that takes `Request` and calls `callback` with `Response` (this call may not happen immediately), it returns `Cancellable` (way to cancel the request).  
+
+`RequestPerformer` also specifies `DataEncoder`, that will be used to encode and decode the data.
+
+#### DataEncoder
+
+```Swift
+protocol DataEncoder {
+    
+    func encodeToQueryString(input: SupportedType, to request: inout Request)
+    
+    func encodeToHttpBody(input: SupportedType, to request: inout Request)
+    
+    func encodeCustom(input: SupportedType, to request: inout Request, inputEncoding: InputEncoding)
+    
+    func decode(response: Response<Data>) -> Response<SupportedType>
+}
+```
+
+`DataEncoder` is called by `Fetcher` to encode the input data into `Request` (`encode???`) and to decode data from `Response` (`decode`). 
+
+`encodeToQueryString` and `encodeToHttpBody` are used if `StandardInputEncoding` is used (method selected based on value). 
+
+`encodeCustom` is called if a custom implementation of `InputEncoding` is used and it is not `InputEncodingWithEncoder`. By default calling this method will result in crash (`InputEncoding` is not known).  
+
+#### Implementations
+
+*AlamofireRequestPerformer*
+
+`AlamofireRequestPerformer` is implementation over [Alamofire](https://github.com/Alamofire/Alamofire). By default it uses `AlamofireJsonDataEncoder` (encoder for JSON) as `DataEncoder` but this can be changed in its `init`. 
+
+`AlamofireJsonDataEncoder` knows special type of `InputEncoding` and that is `FormInputEncoding` which supports content type `application/x-www-form-urlencoded`.
+
+### Header
+
+```Swift
+protocol Header: RequestModifier {
+    
+    var name: String { get }
+    var value: String { get }
+}
+```
+
+`Header` is a special type of `RequestModifier`, which represents headers of the request. Registered instances of `Header` are added to `Request` by internal `RequestEnhancer`.
+
+#### Predefined headers
+
+`Headers` gathers all predefined headers. They are represented as nested structs added by extensions. For example this is implementation of `Accept`:
+
+```Swift
+extension Headers {
+    
+    struct Accept: Header {
+        
+        let name = "Accept"
+        
+        let value: String
+        
+        init(value: String) {
+            self.value = value
+        }
+    }
+}
+
+extension Headers.Accept {
+    
+    static let applicationJson = Headers.Accept(value: "application/json")
+    static let textPlain = Headers.Accept(value: "text/plain")
+}
+```
+
+Another headers can be added the same way (as well as concrete values of the headers).
+
+Currently there are these headers: `Accept`, `ContentType`, `Charset`.
+
+`Custom` is a special type of header. It is meant to by used if some header is needed only in one place, so it is not worth it to create a new struct.
```

Comment:
"... is needed in only one place ..."
"... it is not worth creating a new struct."